core/logger/splunk.go

```diff
@@ -0,0 +1,152 @@
+package logger
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type event struct {
+	Time       int64           `json:"time"`                 // epoch time in nano-seconds
+	Host       string          `json:"host"`                 // hostname
+	Source     string          `json:"source,omitempty"`     // optional description of the source of the event; typically the app's name
+	SourceType string          `json:"sourcetype,omitempty"` // optional name of a Splunk parsing configuration; this is usually inferred by Splunk
+	Index      string          `json:"index,omitempty"`      // optional name of the Splunk index to store the event in; not required if the token has a default index set in Splunk
+	Event      json.RawMessage `json:"event"`                // zap json logger
+}
+
+type splunkSink struct {
+	HTTPClient    httpClient
+	nowFn         func() int64
+	URL           string // URL to Splunk HTTP Event Collector
+	Hostname      string // Hostname is the human-readable machine name sending logs
+	Token         string // Splunk HTTP Event Collector token
+	Source        string // Splunk source field value, description of the source of the event
+	SourceType    string // Splunk source type, optional name of a sourcetype field value
+	Index         string // Splunk Index, optional name of the Splunk index to store the event in
+	SkipTLSVerify bool   // Skip verifying the certificate of the HTTP Event Collector
+	events        [][]byte
+	sync.Mutex
+}
+
+func envVarOrDefault(key string, defaultValue string) string {
+	envVar := os.Getenv(key)
+	if envVar == "" {
+		return defaultValue
+	}
+	return envVar
+}
+
+func initialize(s *splunkSink) *splunkSink {
+	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: s.SkipTLSVerify}}
+	s.HTTPClient = &http.Client{Timeout: time.Second * 20, Transport: tr}
+
+	if s.Hostname == "" {
+		s.Hostname, _ = os.Hostname()
+	}
+	if s.nowFn == nil {
+		s.nowFn = func() int64 {
+			return time.Now().UnixNano()
+		}
+	}
+
+	return s
+}
+
+func newSplunkSink() (*splunkSink, error) {
+	s := &splunkSink{}
+	s.URL = os.Getenv("SPLUNK_URL")
+	s.Source = envVarOrDefault("SPLUNK_SOURCE", "chainlink")
+	s.SourceType = envVarOrDefault("SPLUNK_SOURCETYPE", "chainlink")
+	s.Index = os.Getenv("SPLUNK_INDEX")
+	s.Hostname = os.Getenv("SPLUNK_HOSTNAME")
+	s.Token = os.Getenv("SPLUNK_TOKEN")
+	s.SkipTLSVerify = "true" == os.Getenv("SPLUNK_SKIP_TLS_VERIFY")
```

Comment:
Are these standard for specifying splunk details? It might be better to have these read from the `config` instead.

Reply:
These are. They can definitely be read from your config object instead, especially to set defaults. I just wanted to keep the PR on the neat side and avoid spamming your config object.

Reply:
I will rework that bit.

Reply:
Turns out they cannot be read from the config object, since they go to the splunk sink which is initialized in the init() method of the logger package. Maybe there's a lifecycle change there to be looked at, for now, we have to read the env vars directly.